src/storage-controller/src/lib.rs

```diff
@@ -840,18 +840,33 @@ where
         }
 
         // Register the tables all in one batch.
-        //
-        // We cannot register tables at the table worker when in read-only mode.
-        // When coming out of read-only mode (via `allow_writes()`), we will
-        // register all tables that are known by that point.
         if !table_registers.is_empty() {
             let register_ts = register_ts
                 .expect("caller should have provided a register_ts when creating a table");
 
-            self.persist_table_worker
-                .register(register_ts, table_registers)
-                .await
-                .expect("table worker unexpectedly shut down");
+            if self.read_only {
+                // In read-only mode, we use a special read-only table worker
+                // that allows writing to migrated tables and will continually
+                // bump their shard upper so that it tracks the txn shard upper.
+                // We do this, so that they remain readable at a recent
+                // timestamp, which in turn allows dataflows that depend on them
+                // to (re-)hydrate.
+                //
+                // We only want to register migrated tables, though, and leave
+                // existing tables out/never write to them in read-only mode.
+                table_registers
+                    .retain(|(id, _write_handle)| migrated_storage_collections.contains(id));
+
+                self.persist_table_worker
+                    .register(register_ts, table_registers)
+                    .await
+                    .expect("table worker unexpectedly shut down");
```

Comment:
It seems mildly weird to duplicate the call to `register` when it's identical in both blocks, but I'm assuming you wrote it this way on purpose? Maybe to make it more clear how to evolve this code when we rip out the current approach to built-in migrations?